fieldservice_stage_server_action/models/fsm_order.py

```diff
@@ -0,0 +1,31 @@
+# Copyright 2022 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, models
+
+
+class FSMOrder(models.Model):
+    _inherit = "fsm.order"
+
+    @api.model
+    def create(self, vals):
+        res = super(FSMOrder, self).create(vals)
+        action = self.env["fsm.stage"].browse(vals.get("stage_id")).action_id
+        if action:
+            context = {
+                "active_model": self._name,
+                "active_ids": [res.id],
+            }
+            action.with_context(**context).run()
+        return res
+
+    def write(self, vals):
+        res = super().write(vals)
+        action = self.env["fsm.stage"].browse(vals.get("stage_id")).action_id
```

Comment:
```suggestion
        for record in self:
            action = self.env["fsm.stage"].browse(record.stage_id).action_id
```

Reply:
Actually I think currently it's better

Currently, if `stage_id` is not in `vals`, `vals.get("stage_id")` would be `None`.
And thus `action = self.env["fsm.stage"].browse(None).action_id` would be an empty recordset.

This is expected, as we don't want to trigger any action if `stage_id` wasn't updated.

----

If we don't read `stage_id` from vals, you'd be triggering the action for any write. That is, even if some other random field was updated, and not the stage. This is not desired.


----

I do think, though, this piece of code is a bit difficult to read due to the fact that you rely on `vals.get("stage_id")` to be `None` if it's missing, and that `.browse(None)` returns an empty recordset.

I would prefer somethingl like this instead, to make it more explicit:

```py
    def write(self, vals):
        res = super().write(vals)
        action = self.env["fsm.stage"].browse(vals["stage_id"]).action_id if "stage_id" in vals else None
```

or

```py
    def write(self, vals):
        res = super().write(vals)
        if "stage_id" in vals:
            action = self.env["fsm.stage"].browse(vals.["stage_id"]).action_id
            # ...
```